source/simulator/helper_functions.cc

```diff
@@ -1694,72 +1686,82 @@ namespace aspect
+
+    using VectorType = Vector<double>;
+    SUNDIALS::ARKode<VectorType>::AdditionalData data (0.0, time_step, 0.001*time_step, 0.1*time_step, 1.e-6*time_step);
+    SUNDIALS::ARKode<VectorType> ode(data);
+
     for (const auto &cell : dof_handler.active_cell_iterators())
       if (cell->is_locally_owned())
         {
           fe_values_C.reinit (cell);
           in_C.reinit(fe_values_C, cell, introspection, solution);
 
-          if (temperature_and_composition_use_same_fe == false)
-            {
-              fe_values_T.reinit (cell);
-              in_T.reinit(fe_values_T, cell, introspection, solution);
-            }
-
           std::vector<std::vector<double>> accumulated_reactions_C (quadrature_C.size(),std::vector<double> (introspection.n_compositional_fields));
           std::vector<double> accumulated_reactions_T (quadrature_T.size());
 
-          // Make the reaction time steps: We have to update the values of compositional fields and the temperature.
-          // Because temperature and composition might use different finite elements, we loop through their elements
-          // separately, and update the temperature and the compositions for both.
-          // We can reuse the same material model inputs and outputs structure for each reaction time step.
-          // We store the computed updates to temperature and composition in a separate (accumulated_reactions) vector,
-          // so that we can later copy it over to the solution vector.
-          for (unsigned int i=0; i<number_of_reaction_steps; ++i)
-            {
-              // Loop over composition element
-              material_model->fill_additional_material_model_inputs(in_C, solution, fe_values_C, introspection);
-
-              material_model->evaluate(in_C, out_C);
-              heating_model_manager.evaluate(in_C, out_C, heating_model_outputs_C);
-
-              for (unsigned int j=0; j<dof_handler.get_fe().base_element(introspection.base_elements.compositional_fields).dofs_per_cell; ++j)
+          std::vector<std::vector<double>> initial_values_C (quadrature_C.size(),std::vector<double> (introspection.n_compositional_fields));
+          initial_values_C = in_C.composition;
+          std::vector<double> initial_values_T (quadrature_C.size());
+          initial_values_T = in_C.temperature;
+
+          // We have to store all values of the temperature and composition fields in one long vector
+          const unsigned int n_q_points = dof_handler.get_fe().base_element(introspection.base_elements.compositional_fields).dofs_per_cell;
+          const unsigned int n_fields = 1 + introspection.n_compositional_fields;
+          VectorType fields (n_q_points * n_fields);
+
+          for (unsigned int j=0; j<n_q_points; ++j)
+            for (unsigned int f=0; f<n_fields; ++f)
+              if (f==0)
+                fields[j*n_fields+f] = in_C.temperature[j];
+              else
+                fields[j*n_fields+f] = in_C.composition[j][f-1];
```

Comment:
It might be nice to create a lambda function of the form
```
  auto copy_fields_into_one_vector = [n_q_points,n_fields](const auto &in, auto &fields) {...};
```
that you place somewhere up above, and then you here just have to do
```
  copy_fields_into_one_vector (in_C, fields);
```